### PR TITLE
[Feature] Service instance dependency

### PR DESCRIPTION
**Pull request by arugal**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

- Related issues
#3829
___
### New feature or improvement

+ Storage ```ServiceInstanceRelationServerSideMetrics``` and ```ServiceInstanceRelationClientSideMetrics```
+ Add and update  ```ServiceInstanceInventory#mappingServiceInstanceId```
+ Add ```getServiceInstanceTopology(clientServiceId: ID!, serverServiceId: ID!, duration: Duration!): ServiceInstanceTopology```  to ```TopologyQuery```

___
Sorry, because I deleted the fork repository branch by mistake, I can only open a new PR. 

Ref #3950

### CODE REVIEW

**Comment by arugal** [comment]
works like this: 
service

instance

**Comment by wu-sheng** [comment]
Generally, it is good. Please add test in e2e.

File: oap-server/server-bootstrap/src/main/resources/official_analysis.oal [hunk 1]
```diff
@@ -53,6 +53,24 @@ service_relation_server_p75 = from(ServiceRelation.latency).filter(detectPoint =
...
+service_instance_relation_server_p99 = from(ServiceInstanceRelation.latency).filter(detectPoint == DetectPoint.SERVER).p99(10);
+service_instance_relation_client_p95 = from(ServiceInstanceRelation.latency).filter(detectPoint == DetectPoint.CLIENT).p95(10);
+service_instance_relation_server_p95 = from(ServiceInstanceRelation.latency).filter(detectPoint == DetectPoint.SERVER).p95(10);
+service_instance_relation_client_p90 = from(ServiceInstanceRelation.latency).filter(detectPoint == DetectPoint.CLIENT).p90(10);
+service_instance_relation_server_p90 = from(ServiceInstanceRelation.latency).filter(detectPoint == DetectPoint.SERVER).p90(10);
+service_instance_relation_client_p75 = from(ServiceInstanceRelation.latency).filter(detectPoint == DetectPoint.CLIENT).p75(10);
+service_instance_relation_server_p75 = from(ServiceInstanceRelation.latency).filter(detectPoint == DetectPoint.SERVER).p75(10);
+service_instance_relation_client_p50 = from(ServiceInstanceRelation.latency).filter(detectPoint == DetectPoint.CLIENT).p50(10);
+service_instance_relation_server_p50 = from(ServiceInstanceRelation.latency).filter(detectPoint == DetectPoint.SERVER).p50(10);
```
Comment (wu-sheng) [inline oap-server/server-bootstrap/src/main/resources/official_analysis.oal]: Just FYI @Fine0830 All these metrics will be available for instance relationship metrics. You will be able to get Instance relationship by giving two service IDs through this protocol, https://github.com/apache/skywalking-query-protocol/blob/master/topology.graphqls#L90

File: test/e2e/e2e-base/src/main/resources/instanceTopo.gql [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                id
+                name
+                type
+                isReal
+            }
+            calls {
+                id
+                source
+                detectPoints
+                target
```
Comment (wu-sheng) [inline test/e2e/e2e-base/src/main/resources/instanceTopo.gql]: You should add `serviceId` and `serviceName` here too

**Comment by arugal** [comment]
From the latest query protocol.
image

**Comment by arugal** [comment]
E2e cluster is not yet complete.

**Comment by wu-sheng** [comment]
@Fine0830 This PR is closing to be merged, once it is done. We could work on the UI side to see the instances relationship(add a button on the line metrics tab, then open a box to show the relationship of instances)

**Comment by wu-sheng** [comment]
Let's add a new e2e cluster test without a gateway. We could check the service instance relationship in a more widely used case.

File: test/e2e/e2e-ttl/e2e-ttl-es/src/docker/ttl_official_analysis.oal [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Service scope metrics
+service_p99 = from(Service.latency).p99(10);
```
Comment (wu-sheng) [inline test/e2e/e2e-ttl/e2e-ttl-es/src/docker/ttl_official_analysis.oal]: Add comments here about why and TODO? Link a new issue?

Comment (arugal) [reply test/e2e/e2e-ttl/e2e-ttl-es/src/docker/ttl_official_analysis.oal]: done

**Comment by wu-sheng** [comment]
@kezhenxu94 @dmsolr @JaredTan95 Please review this PR.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/register/ServiceInstanceInventory.java [hunk 4]
```diff
@@ -62,9 +64,13 @@
     @Setter @Getter @Column(columnName = SERVICE_ID) private int serviceId;
     @Setter @Getter @Column(columnName = IS_ADDRESS) private int isAddress;
     @Setter @Getter @Column(columnName = ADDRESS_ID) private int addressId;
+    @Setter(AccessLevel.PRIVATE) @Getter(AccessLevel.PACKAGE) @Column(columnName = NODE_TYPE) private int nodeType;
+    @Setter @Getter @Column(columnName = MAPPING_SERVICE_ID) private int mappingServiceInstanceId;
     @Getter(AccessLevel.PRIVATE) @Column(columnName = PROPERTIES) private String prop;
     @Getter private JsonObject properties;
 
+    @Setter @Getter private boolean resetServiceMapping = false;
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/register/ServiceInstanceInventory.java]: This field should be `resetServiceInstanceMapping`.

Comment (arugal) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/register/ServiceInstanceInventory.java]: done

**Pull request merged**